// @ionic/cli-framework/utils/format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const os = require("os");
const path = require("path");
const chalk_1 = require("chalk");
const sliceAnsi = require("slice-ansi");
const stringWidthModule = require("string-width");
const stripAnsiModule = require("strip-ansi");
const untildify = require("untildify");
const wrapAnsi = require("wrap-ansi");
exports.stringWidth = stringWidthModule;
exports.stripAnsi = stripAnsiModule;
const MIN_TTY_WIDTH = 80;
const MAX_TTY_WIDTH = 120;
exports.TTY_WIDTH = process.stdout.columns ? Math.max(MIN_TTY_WIDTH, Math.min(process.stdout.columns, MAX_TTY_WIDTH)) : Infinity;
function prettyPath(p) {
    p = expandPath(p);
    const cwd = process.cwd();
    const d = path.dirname(p);
    const h = os.homedir();
    const distanceFromCwd = Math.abs(d.split(path.sep).length - cwd.split(path.sep).length);
    if (cwd === d) {
        return '.' + path.sep + path.basename(p);
    }
    else if (d.startsWith(cwd)) {
        return '.' + path.sep + p.substring(cwd.length + 1);
    }
    else if (distanceFromCwd <= 2) {
        const rel = path.relative(cwd, p);
        return rel ? rel : '.';
    }
    else if (p === h) {
        return '~';
    }
    else if (p.indexOf(h) === 0) {
        return '~' + path.sep + p.substring(h.length + 1);
    }
    return p;
}
exports.prettyPath = prettyPath;
function expandPath(p) {
    return path.resolve(untildify(p));
}
exports.expandPath = expandPath;
function indent(n = 4) {
    return new Array(n).fill(' ').join('');
}
exports.indent = indent;
function wordWrap(msg, { width = exports.TTY_WIDTH, indentation = 0, append = '' }) {
    return wrapAnsi(msg, width - indentation - append.length, { trim: false }).split('\n').join(`${append}\n${indent(indentation)}`);
}
exports.wordWrap = wordWrap;
function generateFillSpaceStringList(list, optimalLength = 1, fillCharacter = ' ') {
    if (optimalLength < 2) {
        optimalLength = 2;
    }
    const longestItem = Math.max(...list.map(item => exports.stringWidth(item)));
    const fullLength = longestItem > optimalLength ? longestItem + 1 : optimalLength;
    const fullLengthString = Array(fullLength).fill(fillCharacter).join('');
    return list.map(item => sliceAnsi(fullLengthString, 0, fullLength - exports.stringWidth(item)));
}
exports.generateFillSpaceStringList = generateFillSpaceStringList;
function columnar(rows, options = {}) {
    if (!options.hsep) {
        options.hsep = chalk_1.default.dim('-');
    }
    if (!options.vsep) {
        options.vsep = chalk_1.default.dim('|');
    }
    const includeHeaders = options.columnHeaders ? true : false;
    if (!rows[0]) {
        return '';
    }
    const columnCount = options.columnHeaders ? options.columnHeaders.length : rows[0].length;
    const columns = options.columnHeaders ?
        options.columnHeaders.map(header => [chalk_1.default.bold(header)]) :
        new Array(columnCount).fill([]).map(() => []);
    for (const row of rows) {
        for (const i in row) {
            if (columns[i]) {
                columns[i].push(row[i]);
            }
        }
    }
    const paddedColumns = columns.map((col, i) => {
        if (i < columnCount - 1) {
            const spaceCol = generateFillSpaceStringList(col);
            return col.map((cell, i) => `${cell}${spaceCol[i]}${options.vsep} `);
        }
        else {
            return col;
        }
    });
    let longestRowLength = 0;
    const singleColumn = paddedColumns.reduce((a, b) => {
        return a.map((_, i) => {
            const r = a[i] + b[i];
            longestRowLength = Math.max(longestRowLength, exports.stringWidth(r));
            return r;
        });
    });
    if (includeHeaders) {
        singleColumn.splice(1, 0, options.hsep.repeat(longestRowLength));
    }
    return singleColumn.join('\n');
}
exports.columnar = columnar;
