// @ionic/cli-utils/lib/project/angular/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = require("path");
const chalk_1 = require("chalk");
const Debug = require("debug");
const lodash = require("lodash");
const format_1 = require("@ionic/cli-framework/utils/format");
const fs_1 = require("@ionic/cli-framework/utils/fs");
const npm_1 = require("@ionic/cli-framework/utils/npm");
const _1 = require("../");
const utils_1 = require("./utils");
const debug = Debug('ionic:cli-utils:lib:project:angular');
class Project extends _1.BaseProject {
    constructor() {
        super(...arguments);
        this.type = 'angular';
    }
    getInfo() {
        const _super = name => super[name];
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const [ionicAngularVersion, ionicCoreVersion, angularCLIVersion,] = yield Promise.all([
                this.getFrameworkVersion(),
                this.getCoreVersion(),
                this.getAngularCLIVersion(),
            ]);
            return [
                ...(yield _super("getInfo").call(this)),
                { type: 'local-packages', key: 'Ionic Framework', value: ionicAngularVersion ? `@ionic/angular ${ionicAngularVersion}` : 'not installed' },
                { type: 'local-packages', key: '@ionic/core', value: ionicCoreVersion ? ionicCoreVersion : 'not installed' },
                { type: 'local-packages', key: '@angular/cli', value: angularCLIVersion ? angularCLIVersion : 'not installed' },
            ];
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                const pkg = yield npm_1.readPackageJsonFile(path.resolve(this.directory, 'package.json'));
                const deps = lodash.assign({}, pkg.dependencies, pkg.devDependencies);
                if (typeof deps['@ionic/angular'] === 'string') {
                    debug(`${chalk_1.default.bold('@ionic/angular')} detected in ${chalk_1.default.bold('package.json')}`);
                    return true;
                }
            }
            catch (e) {
                // ignore
            }
            return false;
        });
    }
    personalize(details) {
        const _super = name => super[name];
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield _super("personalize").call(this, details);
            const { appName, displayName } = details;
            const angularJsonPath = path.resolve(this.directory, utils_1.ANGULAR_CLI_FILE);
            try {
                const angularJson = yield utils_1.readAngularCLIJsonFile(angularJsonPath);
                angularJson.project.name = displayName ? displayName : appName;
                yield fs_1.fsWriteJsonFile(angularJsonPath, angularJson, { encoding: 'utf8' });
            }
            catch (e) {
                this.log.error(`Error with ${chalk_1.default.bold(format_1.prettyPath(angularJsonPath))} file: ${e}`);
            }
        });
    }
    getFrameworkVersion() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const pkgPath = path.resolve(this.directory, 'node_modules', '@ionic', 'angular', 'package.json');
            try {
                const pkg = yield npm_1.readPackageJsonFile(pkgPath);
                return pkg.version;
            }
            catch (e) {
                this.log.error(`Error with ${chalk_1.default.bold(format_1.prettyPath(pkgPath))} file: ${e}`);
            }
        });
    }
    getCoreVersion() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const pkgPath = path.resolve(this.directory, 'node_modules', '@ionic', 'core', 'package.json');
            try {
                const pkg = yield npm_1.readPackageJsonFile(pkgPath);
                return pkg.version;
            }
            catch (e) {
                this.log.error(`Error with ${chalk_1.default.bold(format_1.prettyPath(pkgPath))} file: ${e}`);
            }
        });
    }
    getAngularCLIVersion() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const pkgPath = path.resolve(this.directory, 'node_modules', '@angular', 'cli', 'package.json');
            try {
                const pkg = yield npm_1.readPackageJsonFile(pkgPath);
                return pkg.version;
            }
            catch (e) {
                this.log.error(`Error with ${chalk_1.default.bold(format_1.prettyPath(pkgPath))} file: ${e}`);
            }
        });
    }
}
exports.Project = Project;
