// @ionic/cli-utils/lib/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Exception extends Error {
    constructor(message) {
        super(message);
        this.name = 'Exception';
        this.message = message;
        this.stack = (new Error()).stack || '';
    }
    toString() {
        return `${this.name}: ${this.message}`;
    }
}
exports.Exception = Exception;
class FatalException extends Exception {
    constructor(message = '', exitCode = 1) {
        super(message);
        this.message = message;
        this.exitCode = exitCode;
        this.fatal = true;
    }
}
exports.FatalException = FatalException;
class ShellException extends Exception {
    constructor(message, exitCode = 0) {
        super(message);
        this.message = message;
        this.exitCode = exitCode;
    }
}
exports.ShellException = ShellException;
class ServeException extends Exception {
}
exports.ServeException = ServeException;
class ServeCommandNotFoundException extends ServeException {
}
exports.ServeCommandNotFoundException = ServeCommandNotFoundException;
class SessionException extends Exception {
}
exports.SessionException = SessionException;
class RunnerException extends Exception {
}
exports.RunnerException = RunnerException;
class RunnerNotFoundException extends RunnerException {
}
exports.RunnerNotFoundException = RunnerNotFoundException;
class IntegrationException extends Exception {
}
exports.IntegrationException = IntegrationException;
class IntegrationNotFoundException extends IntegrationException {
}
exports.IntegrationNotFoundException = IntegrationNotFoundException;
class HookException extends Exception {
}
exports.HookException = HookException;
