// @ionic/cli-utils/lib/session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk_1 = require("chalk");
const guards_1 = require("../guards");
const errors_1 = require("./errors");
const http_1 = require("./http");
class BaseSession {
    constructor(config, client, project) {
        this.config = config;
        this.client = client;
        this.project = project;
    }
    isLoggedIn() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const c = yield this.config.load();
            return typeof c.tokens.user === 'string';
        });
    }
    logout() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const c = yield this.config.load();
            c.user = {};
            delete c.tokens.user;
            c.git.setup = false;
        });
    }
    getUserToken() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const c = yield this.config.load();
            if (!c.tokens.user) {
                throw new errors_1.SessionException(`Oops, sorry! You'll need to log in:\n    ${chalk_1.default.green('ionic login')}\n\n` +
                    `You can create a new account by signing up:\n\n    ${chalk_1.default.green('ionic signup')}\n`);
            }
            return c.tokens.user;
        });
    }
}
exports.BaseSession = BaseSession;
class ProSession extends BaseSession {
    login(email, password) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { req } = yield this.client.make('POST', '/login');
            req.send({ email, password, source: 'cli' });
            try {
                const res = yield this.client.do(req);
                if (!guards_1.isLoginResponse(res)) {
                    throw http_1.createFatalAPIFormat(req, res);
                }
                const { token, user } = res.data;
                const c = yield this.config.load();
                const user_id = String(user.id);
                if (c.user.id !== user_id) {
                    yield this.logout();
                }
                c.user.id = user_id;
                c.user.email = email;
                c.tokens.user = token;
            }
            catch (e) {
                if (guards_1.isSuperAgentError(e) && (e.response.status === 401 || e.response.status === 403)) {
                    throw new errors_1.SessionException('Incorrect email or password.');
                }
                throw e;
            }
        });
    }
    tokenLogin(token) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { req } = yield this.client.make('GET', '/users/self');
            req.set('Authorization', `Bearer ${token}`);
            try {
                const res = yield this.client.do(req);
                if (!guards_1.isUserResponse(res)) {
                    throw http_1.createFatalAPIFormat(req, res);
                }
                const user = res.data;
                const c = yield this.config.load();
                const user_id = String(user.id);
                if (c.user.id !== user_id) {
                    yield this.logout();
                }
                c.user.id = user_id;
                c.user.email = user.email;
                c.tokens.user = token;
            }
            catch (e) {
                if (guards_1.isSuperAgentError(e) && (e.response.status === 401 || e.response.status === 403)) {
                    throw new errors_1.SessionException('Invalid auth token.');
                }
                throw e;
            }
        });
    }
}
exports.ProSession = ProSession;
function promptToLogin(env) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { validators } = yield Promise.resolve().then(() => require('@ionic/cli-framework'));
        env.log.msg(`Log into your Ionic account\nIf you don't have one yet, create yours by running: ${chalk_1.default.green(`ionic signup`)}\n`);
        const email = yield env.prompt({
            type: 'input',
            name: 'email',
            message: 'Email:',
            validate: v => validators.email(v),
        });
        const password = yield env.prompt({
            type: 'password',
            name: 'password',
            message: 'Password:',
        });
        yield env.session.login(email, password);
    });
}
exports.promptToLogin = promptToLogin;
