// @ionic/cli-utils/lib/doctor/ailments/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = require("path");
const chalk_1 = require("chalk");
const semver = require("semver");
const fs_1 = require("@ionic/cli-framework/utils/fs");
const npm_1 = require("@ionic/cli-framework/utils/npm");
const app_1 = require("../../app");
const git_1 = require("../../git");
const npm_2 = require("../../utils/npm");
const project_1 = require("../../integrations/cordova/project");
const config_1 = require("../../integrations/cordova/config");
const base_1 = require("./base");
tslib_1.__exportStar(require("./base"), exports);
function registerAilments(registry, deps) {
    registry.register(new NpmInstalledLocally(deps));
    registry.register(new IonicCLIInstalledLocally(deps));
    registry.register(new GitNotUsed(deps));
    registry.register(new GitConfigInvalid(deps));
    registry.register(new IonicNativeUpdateAvailable(deps));
    registry.register(new IonicNativeMajorUpdateAvailable(deps));
    registry.register(new IonicNativeOldVersionInstalled(deps));
    registry.register(new UnsavedCordovaPlatforms(deps));
    registry.register(new DefaultCordovaBundleIdUsed(deps));
    registry.register(new ViewportFitNotSet(deps));
    registry.register(new CordovaPlatformsCommitted(deps));
}
exports.registerAilments = registerAilments;
class NpmInstalledLocally extends base_1.AutomaticallyTreatableAilment {
    constructor() {
        super(...arguments);
        this.id = 'npm-installed-locally';
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return (`${chalk_1.default.bold('npm')} is installed locally.\n` +
                `${chalk_1.default.bold('npm')} is typically installed globally and may cause some confusion about versions when other CLIs use it.\n`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return fs_1.pathExists(path.join(this.project.directory, 'node_modules', 'npm'));
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = yield this.config.load();
            const { npmClient } = config;
            const [manager, ...managerArgs] = yield npm_2.pkgManagerArgs({ npmClient, shell: this.shell }, { command: 'uninstall', pkg: 'npm' });
            return [
                {
                    name: `Run: ${chalk_1.default.green(manager + ' ' + managerArgs.join(' '))}`,
                    treat: () => tslib_1.__awaiter(this, void 0, void 0, function* () {
                        yield this.shell.run(manager, managerArgs, {});
                    }),
                },
            ];
        });
    }
}
class IonicCLIInstalledLocally extends base_1.AutomaticallyTreatableAilment {
    constructor() {
        super(...arguments);
        this.id = 'ionic-installed-locally';
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return (`The Ionic CLI is installed locally.\n` +
                `While the CLI can run locally, there's no longer a reason to have it installed locally and it may cause some confusion over configuration and versions.\n`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return fs_1.pathExists(path.join(this.project.directory, 'node_modules', 'ionic'));
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = yield this.config.load();
            const { npmClient } = config;
            const [manager, ...managerArgs] = yield npm_2.pkgManagerArgs({ npmClient, shell: this.shell }, { command: 'uninstall', pkg: 'ionic' });
            return [
                {
                    name: `Run: ${chalk_1.default.green(manager + ' ' + managerArgs.join(' '))}`,
                    treat: () => tslib_1.__awaiter(this, void 0, void 0, function* () {
                        yield this.shell.run(manager, managerArgs, {});
                    }),
                },
            ];
        });
    }
}
class GitNotUsed extends base_1.Ailment {
    constructor() {
        super(...arguments);
        this.id = 'git-not-used';
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return (`Git doesn't appear to be in use.\n` +
                `We highly recommend using source control software such as git (${chalk_1.default.bold('https://git-scm.com')}) to track changes in your code throughout time.\n`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!(yield git_1.isRepoInitialized(this.project.directory))) {
                return true;
            }
            const cmdInstalled = yield this.shell.cmdinfo('git', ['--version']);
            if (!cmdInstalled) {
                return true;
            }
            const [revListCount, status] = yield Promise.all([
                this.shell.run('git', ['rev-list', '--count', 'HEAD'], { showCommand: false }),
                this.shell.run('git', ['status', '--porcelain'], { showCommand: false }),
            ]);
            const commitCount = Number(revListCount);
            const changes = Boolean(status);
            return commitCount === 1 && changes;
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return [
                { name: `Download git if you don't have it installed: ${chalk_1.default.bold('https://git-scm.com/downloads')}` },
                { name: `Learn the basics if you're unfamiliar with git: ${chalk_1.default.bold('https://try.github.io')}` },
                { name: `Make your first commit and start tracking code changes! ðŸ˜` },
            ];
        });
    }
}
class GitConfigInvalid extends base_1.Ailment {
    constructor() {
        super(...arguments);
        this.id = 'git-config-invalid';
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const project = yield this.project.load();
            const appId = project.app_id;
            return (`App linked to ${chalk_1.default.bold(appId)} with invalid git configuration.\n` +
                `This app is linked to an app on Ionic (${chalk_1.default.bold(appId)}), but the git configuration is not valid.\n`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const isLoggedIn = yield this.session.isLoggedIn();
            if (!isLoggedIn) {
                return false;
            }
            const project = yield this.project.load();
            const appId = project.app_id;
            if (!appId) {
                return false;
            }
            if (!(yield git_1.isRepoInitialized(this.project.directory))) {
                return false;
            }
            const remote = yield git_1.getIonicRemote({ shell: this.shell }, this.project.directory);
            if (!remote) {
                return true;
            }
            const token = yield this.session.getUserToken();
            const appLoader = new app_1.App(token, this.client);
            const app = yield appLoader.load(appId);
            if (app.repo_url !== remote) {
                return true;
            }
            return false;
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const args = ['git', 'remote'];
            return [
                { name: `Run: ${chalk_1.default.green('ionic ' + args.join(' '))}` },
            ];
        });
    }
}
class IonicNativeUpdateAvailable extends base_1.AutomaticallyTreatableAilment {
    constructor() {
        super(...arguments);
        this.id = 'ionic-native-update-available';
    }
    getVersionPair() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.currentVersion || !this.latestVersion) {
                try {
                    this.currentVersion = (yield npm_1.readPackageJsonFile(path.resolve(this.project.directory, 'node_modules', '@ionic-native', 'core', 'package.json'))).version;
                }
                catch (e) {
                    // Not installed
                }
                this.latestVersion = yield npm_2.pkgLatestVersion('@ionic-native/core');
            }
            if (!this.currentVersion || !this.latestVersion) {
                return ['0.0.0', '0.0.0'];
            }
            return [this.currentVersion, this.latestVersion];
        });
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const [currentVersion, latestVersion] = yield this.getVersionPair();
            return (`Update available for Ionic Native.\n` +
                `An update is available for Ionic Native (${chalk_1.default.cyan(currentVersion)} => ${chalk_1.default.cyan(latestVersion)}).\n`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.project.type !== 'ionic-angular') {
                return false;
            }
            const [currentVersion, latestVersion] = yield this.getVersionPair();
            const diff = semver.diff(currentVersion, latestVersion);
            return diff === 'minor' || diff === 'patch';
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = yield this.config.load();
            const { npmClient } = config;
            const [, latestVersion] = yield this.getVersionPair();
            const modules = yield fs_1.fsReadDir(path.resolve(this.project.directory, 'node_modules', '@ionic-native'));
            return Promise.all(modules.filter(m => m).map((m) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                const [manager, ...managerArgs] = yield npm_2.pkgManagerArgs({ npmClient, shell: this.shell }, { command: 'install', pkg: `@ionic-native/${m}@${latestVersion ? latestVersion : 'latest'}` });
                return {
                    name: `Run: ${chalk_1.default.green(manager + ' ' + managerArgs.join(' '))}`,
                    treat: () => tslib_1.__awaiter(this, void 0, void 0, function* () {
                        yield this.shell.run(manager, managerArgs, {});
                    }),
                };
            })));
        });
    }
}
class IonicNativeMajorUpdateAvailable extends base_1.Ailment {
    constructor() {
        super(...arguments);
        this.id = 'ionic-native-major-update-available';
    }
    getVersionPair() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.currentVersion || !this.latestVersion) {
                try {
                    this.currentVersion = (yield npm_1.readPackageJsonFile(path.resolve(this.project.directory, 'node_modules', '@ionic-native', 'core', 'package.json'))).version;
                }
                catch (e) {
                    // Not installed
                }
                this.latestVersion = yield npm_2.pkgLatestVersion('@ionic-native/core');
            }
            if (!this.currentVersion || !this.latestVersion) {
                return ['0.0.0', '0.0.0'];
            }
            return [this.currentVersion, this.latestVersion];
        });
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const [currentVersion, latestVersion] = yield this.getVersionPair();
            return (`Major update available for Ionic Native.\n` +
                `A major update is available for Ionic Native (${chalk_1.default.cyan(currentVersion)} => ${chalk_1.default.cyan(latestVersion)}).\n`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.project.type !== 'ionic-angular') {
                return false;
            }
            const [currentVersion, latestVersion] = yield this.getVersionPair();
            const diff = semver.diff(currentVersion, latestVersion);
            return diff === 'major';
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = yield this.config.load();
            const { npmClient } = config;
            const [, latestVersion] = yield this.getVersionPair();
            const args = yield npm_2.pkgManagerArgs({ npmClient, shell: this.shell }, { command: 'install', pkg: `@ionic-native/core@${latestVersion ? latestVersion : 'latest'}` });
            return [
                { name: `Visit ${chalk_1.default.bold('https://github.com/ionic-team/ionic-native/releases')}, looking for breaking changes` },
                { name: `Update each ${chalk_1.default.bold('@ionic-native/')} package. For example, ${chalk_1.default.green(args.join(' '))}` },
                { name: `Update your app according to the breaking changes, if any` },
            ];
        });
    }
}
class IonicNativeOldVersionInstalled extends base_1.Ailment {
    constructor() {
        super(...arguments);
        this.id = 'ionic-native-old-version-installed';
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return (`Old version of Ionic Native installed.\n` +
                `Ionic Native ${chalk_1.default.bold('ionic-native')} has been restructured into individual packages under the ${chalk_1.default.bold('@ionic-native/')} namespace to allow for better bundling and faster apps.\n`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return fs_1.pathExists(path.join(this.project.directory, 'node_modules', 'ionic-native'));
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = yield this.config.load();
            const { npmClient } = config;
            const args = yield npm_2.pkgManagerArgs({ npmClient, shell: this.shell }, { command: 'uninstall', pkg: 'ionic-native' });
            return [
                { name: `Run ${chalk_1.default.green(args.join(' '))}` },
                { name: `Refer to ${chalk_1.default.bold('https://ionicframework.com/docs/native')} for installation & usage instructions` },
            ];
        });
    }
}
class UnsavedCordovaPlatforms extends base_1.Ailment {
    constructor() {
        super(...arguments);
        this.id = 'unsaved-cordova-platforms';
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return (`Cordova platforms unsaved.\n` +
                `There are Cordova platforms installed that are not saved in ${chalk_1.default.bold('config.xml')} or ${chalk_1.default.bold('package.json')}. It is good practice to manage Cordova platforms and their versions. See the Cordova docs${chalk_1.default.cyan('[1]')} for more information.\n\n` +
                `${chalk_1.default.cyan('[1]')}: ${chalk_1.default.bold('https://cordova.apache.org/docs/en/latest/platform_plugin_versioning_ref/')}\n`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const project = yield this.project.load();
            if (!project.integrations.cordova) {
                return false;
            }
            const platforms = yield project_1.getPlatforms(this.project.directory);
            const conf = yield config_1.ConfigXml.load(this.project.directory);
            const engines = conf.getPlatformEngines();
            const engineNames = new Set([...engines.map(e => e.name)]);
            const configXmlDiff = platforms.filter(p => !engineNames.has(p));
            return configXmlDiff.length > 0;
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const args = ['cordova', 'platform', 'save'];
            return [
                { name: `Run: ${chalk_1.default.green('ionic ' + args.join(' '))}` },
            ];
        });
    }
}
class DefaultCordovaBundleIdUsed extends base_1.Ailment {
    constructor() {
        super(...arguments);
        this.id = 'default-cordova-bundle-id-used';
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return (`Bundle ID unchanged in ${chalk_1.default.bold('config.xml')}.\n` +
                `The "bundle identifier" is a unique ID (usually written in reverse DNS notation, such as ${chalk_1.default.bold('com.mycompany.MyApp')}) that Cordova uses when compiling the native build of your app. When your app is submitted to the App Store or Play Store, the bundle ID can't be changed. This issue was detected because this app's bundle ID is ${chalk_1.default.green('"io.ionic.starter"')}, which is the default bundle ID provided after running ${chalk_1.default.green('ionic start')}.`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const project = yield this.project.load();
            if (!project.integrations.cordova) {
                return false;
            }
            const conf = yield config_1.ConfigXml.load(this.project.directory);
            return conf.getBundleId() === 'io.ionic.starter';
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return [
                { name: `Change the ${chalk_1.default.bold('id')} attribute of ${chalk_1.default.bold('<widget>')} (root element) to something other than ${chalk_1.default.green('"io.ionic.starter"')}` },
            ];
        });
    }
}
class ViewportFitNotSet extends base_1.Ailment {
    constructor() {
        super(...arguments);
        this.id = 'viewport-fit-not-set';
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return (`${chalk_1.default.bold('viewport-fit=cover')} not set in ${chalk_1.default.bold('index.html')}\n` +
                `iOS 11 introduces new "safe regions" for webviews, which can throw off component sizing, squish the header into the status bar, letterbox the app on iPhone X, etc. Fixing this issue will ensure the webview takes up the full size of the screen. See ${chalk_1.default.bold('https://blog.ionicframework.com/ios-11-checklist')} for more information.`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const indexHtml = yield fs_1.fsReadFile(path.resolve(yield this.project.getSourceDir(), 'index.html'), { encoding: 'utf8' });
            const m = indexHtml.match(/\<meta.*viewport-fit=cover/);
            return !Boolean(m);
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return [
                { name: `Add ${chalk_1.default.bold('viewport-fit=cover')} to the ${chalk_1.default.bold('<meta name="viewport">')} tag in your ${chalk_1.default.bold('index.html')} file` },
            ];
        });
    }
}
class CordovaPlatformsCommitted extends base_1.Ailment {
    constructor() {
        super(...arguments);
        this.id = 'cordova-platforms-committed';
    }
    getMessage() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return (`Cordova ${chalk_1.default.bold('platforms/')} directory is committed to git.\n` +
                `Cordova considers ${chalk_1.default.bold('platforms/')} and ${chalk_1.default.bold('plugins/')} build artifacts${chalk_1.default.cyan('[1]')}, and routinely overwrites files.\n\n` +
                `While committing these files might be necessary for some projects${chalk_1.default.cyan('[2]')}, generally platforms should be configured using ${chalk_1.default.bold('config.xml')} and Cordova hooks${chalk_1.default.cyan('[3]')} so that your project is more portable and SDK updates are easier.\n\n` +
                `${chalk_1.default.cyan('[1]')}: ${chalk_1.default.bold('https://cordova.apache.org/docs/en/latest/reference/cordova-cli/#version-control')}\n` +
                `${chalk_1.default.cyan('[2]')}: ${chalk_1.default.bold('https://cordova.apache.org/docs/en/latest/reference/cordova-cli/#platforms')}\n` +
                `${chalk_1.default.cyan('[3]')}: ${chalk_1.default.bold('https://cordova.apache.org/docs/en/latest/guide/appdev/hooks/index.html')}\n\n` +
                `${chalk_1.default.yellow(`${chalk_1.default.bold('WARNING')}: Attempting to fix this could be dangerous. Only proceed if you're sure you haven't made manual modifications to these files.`)}\n`).trim();
        });
    }
    detected() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!(yield git_1.isRepoInitialized(this.project.directory))) {
                return false;
            }
            const cmdInstalled = yield this.shell.cmdinfo('git', ['--version']);
            if (!cmdInstalled) {
                return false;
            }
            const files = (yield this.shell.output('git', ['ls-tree', '--name-only', 'HEAD'], { showCommand: false })).split('\n');
            return files.includes('platforms'); // TODO
        });
    }
    getTreatmentSteps() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return [
                { name: `Remove ${chalk_1.default.bold('platforms/')} from source control: ${chalk_1.default.green('git rm -rf platforms/')} and ${chalk_1.default.green('git commit')}` },
                { name: `Make sure the ${chalk_1.default.bold('platforms/')} directory has been removed: ${chalk_1.default.green('rm -rf platforms/')}` },
                { name: `Allow Cordova to repopulate your platforms: ${chalk_1.default.green('ionic cordova prepare')}` },
            ];
        });
    }
}
