// @ionic/cli-utils/lib/project/ionic-angular/app-scripts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path = require("path");
const constants_1 = require("../../../constants");
function importAppScripts(env) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const appScriptsPath = path.resolve(env.project.directory, 'node_modules', '@ionic', 'app-scripts'); // TODO
        return require(appScriptsPath);
    });
}
exports.importAppScripts = importAppScripts;
exports.APP_SCRIPTS_OPTIONS = [
    {
        name: 'prod',
        description: 'Build the application for production',
        type: Boolean,
        groups: [constants_1.OptionGroup.AppScripts],
        hint: 'app-scripts',
    },
    {
        name: 'aot',
        description: 'Perform ahead-of-time compilation for this build',
        type: Boolean,
        groups: [constants_1.OptionGroup.Advanced, constants_1.OptionGroup.AppScripts],
        hint: 'app-scripts',
    },
    {
        name: 'minifyjs',
        description: 'Minify JS for this build',
        type: Boolean,
        groups: [constants_1.OptionGroup.Advanced, constants_1.OptionGroup.AppScripts],
        hint: 'app-scripts',
    },
    {
        name: 'minifycss',
        description: 'Minify CSS for this build',
        type: Boolean,
        groups: [constants_1.OptionGroup.Advanced, constants_1.OptionGroup.AppScripts],
        hint: 'app-scripts',
    },
    {
        name: 'optimizejs',
        description: 'Perform JS optimizations for this build',
        type: Boolean,
        groups: [constants_1.OptionGroup.Advanced, constants_1.OptionGroup.AppScripts],
        hint: 'app-scripts',
    },
    {
        name: 'env',
        description: '',
        groups: [constants_1.OptionGroup.Hidden, constants_1.OptionGroup.AppScripts],
        hint: 'app-scripts',
    },
];
