// @ionic/cli-utils/lib/project/ionic-angular/serve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk_1 = require("chalk");
const Debug = require("debug");
const through2 = require("through2");
const split2 = require("split2");
const cli_framework_1 = require("@ionic/cli-framework");
const string_1 = require("@ionic/cli-framework/utils/string");
const constants_1 = require("../../../constants");
const errors_1 = require("../../errors");
const serve_1 = require("../../serve");
const _1 = require("../");
const app_scripts_1 = require("./app-scripts");
const debug = Debug('ionic:cli-utils:lib:project:ionic-angular:serve');
exports.DEFAULT_PROGRAM = 'ionic-app-scripts';
exports.DEFAULT_SERVE_SCRIPT_VALUE = `${exports.DEFAULT_PROGRAM} serve`;
const APP_SCRIPTS_SERVE_CONNECTIVITY_TIMEOUT = 20000; // ms
class ServeRunner extends serve_1.ServeRunner {
    getCommandMetadata() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return {
                options: [
                    {
                        name: 'consolelogs',
                        description: 'Print app console logs to Ionic CLI',
                        type: Boolean,
                        aliases: ['c'],
                    },
                    {
                        name: 'serverlogs',
                        description: 'Print dev server logs to Ionic CLI',
                        type: Boolean,
                        aliases: ['s'],
                        groups: [constants_1.OptionGroup.Hidden],
                    },
                    {
                        name: 'livereload-port',
                        description: 'Use specific port for live-reload',
                        default: serve_1.DEFAULT_LIVERELOAD_PORT.toString(),
                        aliases: ['r'],
                        groups: [constants_1.OptionGroup.Advanced],
                    },
                    {
                        name: 'dev-logger-port',
                        description: 'Use specific port for dev server communication',
                        default: serve_1.DEFAULT_DEV_LOGGER_PORT.toString(),
                        groups: [constants_1.OptionGroup.Advanced],
                    },
                    ...app_scripts_1.APP_SCRIPTS_OPTIONS,
                ],
            };
        });
    }
    createOptionsFromCommandLine(inputs, options) {
        const baseOptions = super.createOptionsFromCommandLine(inputs, options);
        const livereloadPort = string_1.str2num(options['livereload-port'], serve_1.DEFAULT_LIVERELOAD_PORT);
        const notificationPort = string_1.str2num(options['dev-logger-port'], serve_1.DEFAULT_DEV_LOGGER_PORT);
        return Object.assign({}, baseOptions, { consolelogs: options['consolelogs'] ? true : false, serverlogs: options['serverlogs'] ? true : false, livereloadPort,
            notificationPort, env: options['env'] ? String(options['env']) : undefined });
    }
    serveProject(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { findClosestOpenPort, isHostConnectable } = yield Promise.resolve().then(() => require('../../utils/network'));
            const [externalIP, availableInterfaces] = yield this.selectExternalIP(options);
            const port = yield findClosestOpenPort(options.port, '0.0.0.0');
            options.port = port;
            const { program } = yield this.serveCommandWrapper(options);
            debug('waiting for connectivity with app-scripts (%dms timeout)', APP_SCRIPTS_SERVE_CONNECTIVITY_TIMEOUT);
            yield isHostConnectable('localhost', port, APP_SCRIPTS_SERVE_CONNECTIVITY_TIMEOUT);
            return {
                custom: program !== exports.DEFAULT_PROGRAM,
                protocol: 'http',
                localAddress: 'localhost',
                externalAddress: externalIP,
                externalNetworkInterfaces: availableInterfaces,
                port,
                externallyAccessible: ![serve_1.BIND_ALL_ADDRESS, ...serve_1.LOCAL_ADDRESSES].includes(externalIP),
            };
        });
    }
    serveCommandWrapper(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { promptToInstallPkg } = yield Promise.resolve().then(() => require('../../utils/npm'));
            try {
                return yield this.servecmd(options);
            }
            catch (e) {
                if (!(e instanceof errors_1.ServeCommandNotFoundException)) {
                    throw e;
                }
                const pkg = '@ionic/app-scripts';
                this.env.log.nl();
                this.env.log.warn(`Looks like ${chalk_1.default.green(pkg)} isn't installed in this project.\n` +
                    `This package is required for ${chalk_1.default.green('ionic serve')} in ${_1.prettyProjectName('angular')} projects.`);
                const installed = yield promptToInstallPkg(this.env, { pkg, saveDev: true });
                if (!installed) {
                    throw new errors_1.FatalException(`${chalk_1.default.green(pkg)} is required for ${chalk_1.default.green('ionic serve')} to work properly.`);
                }
                return this.servecmd(options);
            }
        });
    }
    servecmd(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { pkgManagerArgs } = yield Promise.resolve().then(() => require('../../utils/npm'));
            const { registerShutdownFunction } = yield Promise.resolve().then(() => require('../../process'));
            const config = yield this.env.config.load();
            const pkg = yield this.env.project.loadPackageJson();
            const { npmClient } = config;
            let program = exports.DEFAULT_PROGRAM;
            let args = yield this.serveOptionsToAppScriptsArgs(options);
            const shellOptions = { cwd: this.env.project.directory, env: Object.assign({ FORCE_COLOR: chalk_1.default.enabled ? '1' : '0' }, process.env) };
            debug(`Looking for ${chalk_1.default.cyan(serve_1.SERVE_SCRIPT)} npm script.`);
            if (pkg.scripts && pkg.scripts[serve_1.SERVE_SCRIPT]) {
                if (pkg.scripts[serve_1.SERVE_SCRIPT] === exports.DEFAULT_SERVE_SCRIPT_VALUE) {
                    debug(`Found ${chalk_1.default.cyan(serve_1.SERVE_SCRIPT)}, but it is the default. Not running.`);
                    args = ['serve', ...args];
                }
                else {
                    debug(`Invoking ${chalk_1.default.cyan(serve_1.SERVE_SCRIPT)} npm script.`);
                    const [pkgManager, ...pkgArgs] = yield pkgManagerArgs({ npmClient, shell: this.env.shell }, { command: 'run', script: serve_1.SERVE_SCRIPT, scriptArgs: [...args] });
                    program = pkgManager;
                    args = pkgArgs;
                }
            }
            else {
                args = ['serve', ...args];
            }
            const p = yield this.env.shell.spawn(program, args, shellOptions);
            return new Promise((resolve, reject) => {
                p.on('error', (err) => {
                    if (program === exports.DEFAULT_PROGRAM && err.code === 'ENOENT') {
                        reject(new errors_1.ServeCommandNotFoundException(`${chalk_1.default.bold(exports.DEFAULT_PROGRAM)} command not found.`));
                    }
                    else {
                        reject(err);
                    }
                });
                registerShutdownFunction(() => p.kill());
                const log = this.env.log.clone({ prefix: chalk_1.default.dim(`[${program === exports.DEFAULT_PROGRAM ? 'app-scripts' : program}]`), wrap: false });
                const ws = log.createWriteStream();
                if (program === exports.DEFAULT_PROGRAM) {
                    const stdoutFilter = through2(function (chunk, enc, callback) {
                        const str = chunk.toString();
                        if (str.includes('server running')) {
                            resolve({ program }); // TODO: https://github.com/ionic-team/ionic-app-scripts/pull/1372
                        }
                        else {
                            this.push(chunk);
                        }
                        callback();
                    });
                    p.stdout.pipe(split2()).pipe(stdoutFilter).pipe(ws);
                    p.stderr.pipe(split2()).pipe(ws);
                }
                else {
                    p.stdout.pipe(split2()).pipe(ws);
                    p.stderr.pipe(split2()).pipe(ws);
                    resolve({ program });
                }
            });
        });
    }
    serveOptionsToAppScriptsArgs(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const args = {
                _: [],
                address: options.address,
                port: String(options.port),
                livereloadPort: String(options.livereloadPort),
                devLoggerPort: String(options.notificationPort),
                consolelogs: options.consolelogs,
                serverlogs: options.serverlogs,
                nobrowser: true,
                nolivereload: !options.livereload,
                noproxy: !options.proxy,
                iscordovaserve: options.engine === 'cordova',
                platform: options.platform,
                target: options.engine === 'cordova' ? 'cordova' : undefined,
                env: options.env,
            };
            return [...cli_framework_1.unparseArgs(args, { useEquals: false }), ...options['--']];
        });
    }
}
exports.ServeRunner = ServeRunner;
